### DISCUSSION TITLE
FileMaker objects Metadata cache

### DISCUSSION THREAD

**Discussion by marcel-more** [fmIDE/fmIDE#20] [Ideas]
To make more sense of a scripted development environment where fmIDE plays a key role in organizing and dispatching dev tasks, it would be convenient to have access to all objects metadata to manage and coordinate access and allow for automatic context related decisions when modifying objects and code.

At his moment where FileMakers' own design functions are limited the only suitable way is to rely on DDR oder 'Save as XML' to  get full access to all objects in form a standardized format. Despite the fact that updating XML information needs to be done regularly it could even mean that accessing and querying the XML-files for a larger solution is a real performance bottleneck.

So it would be great to have some kind of high performance cache, where all object information of the solutions XML files are stored in a way that could be retrieved by simple object queries. As fmIDE for itself already brings its own query standard for any FileMaker object it would be a natural extension to find some method of routing the existing object URLs to some query standard to retrieve the objects meta data. Think of it as some kind of XPATH query method to describe any objects location inside the given solution and even allow access to sister, child or ancestor elements.

Perhaps it could be possible to import all XML files of a solution into a MongoDB as an object store. This process should include proper translation of all objects inside any XML to some kind of key-value-based mapping. On the other side there should be some API endpoint to allow simple queries in a format that is similar to the fmIDE objects URL format.

This would be a simple approach to extend fmIDEs' possibilities as object metadata could be retrieved inside any fmIDE script logic and then apply customized modifications of any object inside the developers scope.

**Comment by mrwatson-de** [comment]
Wow, Marcel, ... it's great to have you on board … and your creative juices flowing - if not pouring - as mine have been for the last month!

The idea of a MongoDB as a Metadata cache is a good idea, ... the idea of leaving metadata files (DDR/XML) lying around on a fileserver seems to be a huge security risk. With a Mongo DB (or similar) it would be possible to implement some kind of security system to protect your metadata as well as you protect your solutions.

**Comment by mrwatson-de** [comment]
...and of course, MBS plugin has MongoDB functions, so maybe that could be easily integrated (without having to touch an API).

these are *very* exciting times! 👍
